File: 30_GCD/30_GCD.py
# Pyhton program to find GCD of a given number
# GCD - greatest common divisor is the largest common number that completely divides the given two numbers

# method 1 - using iteration
a,b=map(int,input("Enter two space separate numbers :").split(" "))
for i in range(min(a,b)):
    if a%i==0 and b%i==0:
        gcd=i
print(gcd)

# method 2 - using recursion
a,b=map(int,input("Enter two space separate numbers :").split(" "))
def fun1(i):
    if i>min(a,b):
        return print(end="")
    if a%i==0 and b%i==0:
        global gcd
        gcd=i
    fun1(i+1)
fun1(1)
print(gcd,"is the GCD")

# method 3 - using module based approach

def gdc(num1, num2):
    minn = min([num1,num2])
    for i in range(1, minn + 1):
        if (num1%minn and num2%minn):
            return minn

try:
    while(True):
        num1 = int(input("Enter positive integer value : "))
        num2 = int(input("Enter positive integer value : "))
        limit  = 1000
        if (num1 == 0 or num2 == 0):
            print("Enter positive integer value")
        elif (num1 > limit or num2 > limit):
            print("Input too large")
        else:
            result = gdc(num1, num2)
            print(f"GCD = {result}")
    
        choice = input("Do you want to continue the program (y/n) : ")
        if(choice == "n"):
            print("Program finished!")
            break

except ValueError as e:
    print("Invalid Input : {e}")

except ZeroDivisionError as e:
    print(f"Zero division exception : {e}")

except Exception as e:
    print(f"Exception caught : {e}")

File: 30_GCD/30_GCD.dart
import 'dart:io';
import 'dart:math';

// GDC - greates common divisor - dart program to finf the gdc of two numbers

int gcd(int num1, int num2) {
  int minn = min(num1, num2);
  for (int i = 0; i <= minn; i++) {
    if (num1 % i == 0 && num2 % i == 0) {
      break;
    }
  }
  return minn;
}

int eucledianMethod(int num1, int num2) {
  while (num2 > 0) {
    int temp = num1;
    num1 = num2;
    num2 = temp % num2;
  }
  return num1;
}

void main() {
  try {
    while (true) {
      stdout.write("Enter a positive integer value : ");
      int num1 = int.parse(stdin.readLineSync()!);

      stdout.write("Enter a positive integer value : ");
      int num2 = int.parse(stdin.readLineSync()!);

      int maxLimit = 1000;

      if (num1 > maxLimit || num2 > maxLimit) {
        print("Enter a lesser value than 10000");
      } else if (num1 == 0 || num2 == 0) {
        print("Enter a valid integer value : ");
      } else {
        int result = gcd(num1, num2);
        int euclideanResult = eucledianMethod(num1, num2);
        print("GCD = $result");
        print("GCD - $euclideanResult");
      }

      stdout.write("Do you want to continue the program (y/n) : ");
      String choice = stdin.readLineSync()!.toLowerCase();
      if (choice == 'n') {
        print("Program finished!");
        break;
      }
    }
  } on FormatException catch (e) {
    print("Invalid input : $e");
  } on UnsupportedError catch (e) {
    print("Divide by zero exception : $e");
  } on Exception catch (e) {
    print("Exception caught : $e");
  }
}
